### ISSUE TITLE
Clouddriver: stops syncing Docker registry at intermittent intervals

### ISSUE THREAD

**Issue by ahoward-conga** [spinnaker/spinnaker#6231]
### Issue Summary:

Potentially related issue: #4780

Clouddriver will stop syncing with our Docker registry at intermittent intervals. There is no error message logged by Clouddriver, event after setting the log level to `TRACE`.

### Cloud Provider(s):

AWS EKS

### Environment:

- EKS v1.18
- Halyard 1.40.0
- Spinnaker 1.23.3
- Clouddriver 7.0.3

### Feature Area:

Clouddriver

### Additional Details:

When working correctly, Clouddriver will output a info message stating that it is going to start describing the cached items in the registry, and follow up with the amount of ids and tags it grabs:

```
2020-12-14 07:01:10.838  INFO 1 --- [cutionAction-29] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in <docker-registry>/DockerRegistryImageCachingAgent[1/1]
2020-12-14 07:01:13.103  INFO 1 --- [cutionAction-29] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 21174 tagged images in <docker-registry>/DockerRegistryImageCachingAgent[1/1]
2020-12-14 07:01:13.103  INFO 1 --- [cutionAction-29] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 21174 image ids in <docker-registry>/DockerRegistryImageCachingAgent[1/1]
```

However, I've noticed that after Clouddriver stops syncing the registry, the cached tags and ids are not printing, and we only get the message from the describe call:
```
2020-12-13 08:09:57.750  INFO 1 --- [utionAction-116] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in <docker-registry>/DockerRegistryImageCachingAgent[1/1]
```
From this point onwards, Clouddriver will not sync with the registry until it (Clouddriver) has been restarted. Once it has been restarted, it will sync with the registry just fine until this intermittent failure occurs again.

It seems like if there is a problem here with Clouddriver proper, the offending code would be somewhere in the `buildCacheResult` method of the `DockerRegistryImageCachingAgent`: https://github.com/spinnaker/clouddriver/blob/874cef3644c826a1932d9670d59ccb77a19e19a2/clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy#L135

I have also checked connectivity to our Docker registry during these outages, and I can connect to it just fine while this issue occurs with Clouddriver.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by german-muzquiz** [comment]
Did you enable `trackDigests` or `sortTagsByDate` in your docker account configuration? If so, for every tag (21174 in your example) clouddriver will do a network request to the registry. It's possible that those requests are hanging for some reason. 

How much time have you waited for the tags to process? If you query clouddriver's endpoint `/cache/introspection` you should see how long it took for the last execution of `DockerRegistryImageCachingAgent` to complete.

**Comment by ahoward-conga** [comment]
> Did you enable `trackDigests` or `sortTagsByDate` in your docker account configuration?

We do not use `trackDigests` or `sortTagsByDate` in our config.

> How much time have you waited for the tags to process?

Basically, for days. Sometimes this would manifest itself over a weekend and when people started pushing new Docker tags after getting back they would notice they weren't showing up in the UI. Looking back at the logs, we could see that the caching agent would run `Describing items in <docker-registry>` but it would never actually cache anything at all and would stop running any further describes after that point. It would never start running a describe for that registry again until Clouddriver was restarted.

However, this may be indicative of a broader caching issue in general, specifically with cache resources. We noticed that one of our clusters stopped showing up at all in the UI (meaning in the Infrastructure tab the cluster's deployments, replicasets, and pods were mysteriously missing). After searching the Issues here I stumbled across https://github.com/spinnaker/spinnaker/issues/5611 where it was suggested to up the resource limits for Redis.

We deployed this about a week ago, and the cluster has been available in the UI since. A couple of days ago I shut off the CronJob we were running to automatically restart Clouddriver in the early mornings and the caching agent has connected with the Docker registry in question flawlessly ever since. But, because of the intermittent nature of this issue (it would happen between every 3 to 6 days or so), I'm not certain that it was just cache resource limitations just yet.

**Comment by german-muzquiz** [comment]
It's possible that you're running into scalability problems as you said, however I would expect the docker caching agent to take several minutes/hours to complete, not hang for days. Anyway the output of `/cache/introspection` would still be useful because it prints the "last run" statistics for all caching agents.

When the issue happens again, it would be helpful if you take a thread dump by running this command from inside clouddriver pod:

```
kill -3 1
```

Then we could find the thread used by the agent from the log 

```
2020-12-14 07:01:10.838  INFO 1 --- [cutionAction-29] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in <docker-registry>/DockerRegistryImageCachingAgent[1/1]
``` 

And see if it's blocked by other thread or waiting for something.

**Comment by ahoward-conga** [comment]
I hit the `/cache/introspection` endpoint a bunch of times and got some statistics for the `lastExecutionDurationMs` on the Docker registry.

Average was around `10933ms` with a maximum of `23959ms` and a minimum time of `9377ms`.

These stats were garnered with the Redis instance scaled up.

We use the default cache interval of 30 seconds specified [here](https://spinnaker.io/reference/halyard/commands/#hal-config-provider-docker-registry-account-add). Considering that the upper bound is around 24 seconds, is it possible that if we were to exceed the cache polling time multiple times in a row that it would cause some sort of thread lock? Just spitballing here.

**Comment by german-muzquiz** [comment]
The 30 seconds start counting from the time it finishes, so the agent would usually poll every (30s + Last agent duration + Time to acquire a lock among all agents).

Also the log message indicates that the agent already started but never finsihed, so an explanation could be the agent thread being locked by some reason. Docker agent seems to use the `ForkJoinPool.commonPool` thread pool for processing image tags, and that pool may also be used by other processes like `kubectl get namespace` calls, so if there are hanging kubectl calls (like for an offline cluster) that may block the docker agent from acquiring a thread for processing the tags. Only with a thread dump you could see who blocks the agent thread.

**Comment by ahoward-conga** [comment]
It seems that upping the resource limits on our Redis instance has solved this issue for us for now. I'm going to consider this issue closed and I can create a new ticket or re-open if I encounter any more issues linked to the caching agent and the Docker registry.

Thank you for your help @german-muzquiz! I will be sure to inform you further and get a thread dump if we run into this again.

**Comment by ahoward-conga** [comment]
So we finally hit the issue again. It looks like upping the cache resources just delayed the manifestation of the problem instead of rectifying it.

Here is the output of the `/cache/introspection` endpoint for the Docker registry. 

```
{
  "details": {},
  "id": "<docker-registry>/DockerRegistryImageCachingAgent[1/1]",
  "lastExecutionDurationMs": 11237,
  "lastExecutionStartDate": "2021-02-23 10:19:32.032",
  "lastExecutionStartMs": 1614075572032,
  "provider": "com.netflix.spinnaker.clouddriver.docker.registry.provider.DockerRegistryProvider",
  "totalAdditions": 47374,
  "totalEvictions": 0
}
```

Also, running `ps` on the Clouddriver pod didn't show `kubectl` being stuck running on anything.

```
1 spinnake  4d11 java -Djava.security.egd......
13088 spinnake  0:00 ps -elf
17512 spinnake  0:00 bash
```

Running `kill -3 1` did absolutely nothing. The `java` command running Clouddriver didn't exit at all, and I didn't get a core dump. It seems like the process may be locked entirely and is not responding to `SIGQUIT`.

@german-muzquiz Do you have any other tips or tricks that I could use to troubleshoot?

**Comment by german-muzquiz** [comment]
Running `kill -3 1` will print the threaddump to clouddriver logs, it doesn't stop the java process and doesn't print anything to the console where the command runs.

According to `/cache/introspection` it took 11 seconds to run successfully the last time, which doesn't seem too bad considering the amount of tags being cached. One test you can try is increasing the value of `cacheThreads` in the docker account configuration (defaults to 1). This will create different caching agent instances like `<docker-registry>/DockerRegistryImageCachingAgent[1/N]`, `<docker-registry>/DockerRegistryImageCachingAgent[2/N]` and so on, each caching different repositories. In this way you could isolate if the problem only happens consistently in one caching agent (a set of specific repositories) or if it presents in all agents, which will also make it more visible in a thread dump if all threads become locked.

This is assuming that the agents start successfully (there's the log `Describing items in...` without `Caching xxx tagged images...`). If the agents doesn't even start it's a different problem.

**Comment by ahoward-conga** [comment]
> Running `kill -3 1` will print the threaddump to clouddriver logs, it doesn't stop the java process and doesn't print anything to the console where the command runs.

Ah, I see! I was expecting the process to terminate and dump a core file.

I'm currently investigating the core dumps with a few analysis tools. I'm new to Java profiling.

You might have been right on the money with:
> if there are hanging kubectl calls (like for an offline cluster) that may block the docker agent from acquiring a thread for processing the tags

We do have a couple of offline clusters that may be causing this issue.

I'll see if I can get any more information and report back with what I find.

**Comment by ahoward-conga** [comment]
@german-muzquiz 

Just an update here.
I did multiple thread dumps for when this issue arose and I couldn't find anything conclusive. The only thing that I was able to see was that the agent threads were typically waiting, but they didn't report themselves as "blocked".

### Docker registry agent stall 1
![image](https://user-images.githubusercontent.com/51962372/110006585-1231d280-7cd7-11eb-93a5-68978780d105.png)

### Docker registry agent stall 2
![image](https://user-images.githubusercontent.com/51962372/110006628-1eb62b00-7cd7-11eb-9fb4-c4415d268953.png)

### Normal operation
![image](https://user-images.githubusercontent.com/51962372/110006688-31306480-7cd7-11eb-896d-d0e2b7cd4682.png)

Interestingly enough, the `Possible deadlock` in the last image might be an issue with the tool at hand, because that was a capture when everything was running fine!

However, you can see the `unknown` threads in the first image. Those threads were actually blocked according to another tool:
![image](https://user-images.githubusercontent.com/51962372/110007188-aac85280-7cd7-11eb-8e35-a0e0772ccf01.png)

The threads that the jstack.review threads claimed were blocked were the `Common-Cleaner` and `Exec Stream Pumper`, however, not one of the Agent threads :/
![image](https://user-images.githubusercontent.com/51962372/110007312-c4699a00-7cd7-11eb-96a4-fc948ea76367.png)

---

I did recently push a couple of changes in order to see if they might help:

1. I removed some clusters that were part of the configuration but destroyed and no longer available per:
> ...that pool may also be used by other processes like kubectl get namespace calls, so if there are hanging kubectl calls (like for an offline cluster) that may block the docker agent from acquiring a thread for processing the tags...
2. Explicitly add `--cache-threads 2` to all of the `kubernetes accounts` and `docker-registry account` configurations.
3. Upgrade to Spinnaker version `1.23.7` and Halyard to `1.41.0`

Again, I can let you know how that goes.

**Comment by german-muzquiz** [comment]
Thanks for the update. I don't have an answer but can provide more context about how this works: Besides the thread of the docker agent being locked (so waiting to acquire one of the threads from the ForkJoinPool commonPool), maybe the only other thing that could cause this is that the docker caching agent is not started. This can be the case for various reasons, the SQL agent scheduler has a [known issue ](https://github.com/spinnaker/spinnaker/issues/5829) that [should be fixed](https://github.com/spinnaker/clouddriver/pull/5254) in `1.26`, but if you're using redis scheduler that doesn't apply in your case.

The setting `[sql/redis].agent.max-concurrent-agents` controls how many agents can run at the same time, which defaults to 1000 for redis and 100 for sql. A kubernetes account can spawn between 1 to N caching agents depending on account configuration, so if you have a number of accounts of any type that comes close or exceeds the `max-concurrent-agents`, the scheduler will select randomly the next set of agents to run, and it performs this check every second. Probably the docker agent is not included in the batch, but never been included in a random selection that is done every second may not be very likely either.

**Comment by atyutyunnik** [comment]
I am seeing this kind of issue with 1.23.7 as well. So the workaround here is periodically restart clouddriver, right?

**Comment by ahoward-conga** [comment]
@atyutyunnik Correct, although after updating the the `--cache-threads` flag for each of our cluster accounts, I haven't seen this issue for quite some time.

If you are unable to update the flag, and you are running Spinnaker in Kubernetes, you can use the following manifests to achieve an automated restart of Clouddriver. Do note that this will delete the Clouddriver pods which _will cause a brief outage and subsequently some downtime_, so if you are unable to handle having downtime I would suggest trying to tune the `--cache-threads` flag instead.

Just save the file as something like `clouddriver-restart.yaml` and run `kubectl apply --filename clouddriver-restart.yaml`. You may have to adjust and tune some of the values (eg namespace, Clouddriver pod name, cron time, etc...) to your specific configuration.

```
# This will create a CronJob that restarts Clouddriver and the necessary RBAC roles to run the the CronJob.
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  namespace: spinnaker
  name: clouddriver-restart
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
  - list
  - delete
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: clouddriver-restart
  namespace: spinnaker
subjects:
- kind: ServiceAccount
  name: clouddriver-restart
  namespace: spinnaker
roleRef:
  kind: Role
  name: clouddriver-restart
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: clouddriver-restart
  namespace: spinnaker
---
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: clouddriver-restart
  namespace: spinnaker
spec:
  schedule: "0 7 * * *" # At 07:00 UTC every day (03:00 Eastern, 00:00 Pacific)
  jobTemplate:
    spec:
      template:
        spec:
          serviceAccountName: clouddriver-restart
          containers:
          - name: kubectl
            image: bitnami/kubectl:1.18.12
            command: ["/bin/sh", "-c"]
            args: ["kubectl delete pod --selector app.kubernetes.io/name=clouddriver"]
          restartPolicy: OnFailure
```

**Comment by atyutyunnik** [comment]
Alex, thanks! I was thinking of the same forcible restarts. This is a most
elegant solution to an otherwise horrible issue.

**Comment by kskewes** [comment]
Perhaps a `kubectl rollout restart deploy clouddriver` or using the label selector is a way to avoid downtime if you don't have `PodDisruptionBudget` configured.
Be good to understand where the issue is though...

**Comment by atyutyunnik** [comment]
> Perhaps a `kubectl rollout restart deploy clouddriver` or using the label selector is a way to avoid downtime if you don't have `PodDisruptionBudget` configured.
> Be good to understand where the issue is though...

Karl, thank you, the suggested restart is even better since there is no interruption of service anymore. The issue is too elusive but painful nonetheless.

**Comment by ahoward-conga** [comment]
@kskewes Thank you! I will edit my comment with your suggestion.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by hisener** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by hisener** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by Stephan202** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by vijay-sapient** [comment]
Hi @german-muzquiz @ahoward-conga , Could you please let me know if this is resolved in any other versions. am using 1.26.4 and I see exact same problem, Images get detected fine for a week or so and then only once we restart we are able to see images

**Comment by german-muzquiz** [comment]
@vijay-sapient there hasn't been any work to address this issue that I'm aware of. I suggest to use the SQL backend instead of redis, since all new work is targeted at the SQL backend and it is also much better for performance.

**Comment by atyutyunnik** [comment]
> @vijay-sapient there hasn't been any work to address this issue that I'm aware of. I suggest to use the SQL backend instead of redis, since all new work is targeted at the SQL backend and it is also much better for performance.

The same issue is present with the SQL backend. The only thing that helps is a daily restart of cloud driver

**Comment by vijay-sapient** [comment]
HI @german-muzquiz  @atyutyunnik Thanks for your response, Is there anyone who can help to look into or prioritize this work, as this is a very important feature in spinnaker, and restarting would not be an option for long term.

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by Stephan202** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by hisener** [comment]
@spinnakerbot remove-label stale

**Comment by vijay-sapient** [comment]
> > @vijay-sapient there hasn't been any work to address this issue that I'm aware of. I suggest to use the SQL backend instead of redis, since all new work is targeted at the SQL backend and it is also much better for performance.

@atyutyunnik is it possible to run spinnaker without redis? I thought this was a hard dependency. We did migrate to cloudsql but is there a way we can remove redis, if you could point me to the documentation would be really helpful.

Thanks,
Vj

**Comment by philleonard** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by Badbond** [comment]
Can verify this is also still happening in version 1.26.7.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by rickie** [comment]
@spinnakerbot remove-label stale

**Comment by Badbond** [comment]
> Can verify this is also still happening in version 1.26.7.

With the upgrade to 1.29.2, it appears that we _likely_ do not have this issue anymore. Clouddriver has been running for ~5 days now and still syncs with the Docker registry. Only time will tell if it is truly the case.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by Badbond** [comment]
> Only time will tell if it is truly the case.

We have not seen the issue anymore. The only issue we've been seeing is that caching is stopped when trying to onboard a new  Docker repository with more than 1000 images in it. This is due to a safeguard by Igor's polling monitor. See for more information:

- https://kb.armory.io/s/article/Hitting-Igor-s-caching-thresholds
- https://github.com/spinnaker/igor/blob/b76c323e191b8a8acbe825f8a832ddd3f9460fb5/igor-core/src/main/java/com/netflix/spinnaker/igor/IgorConfigurationProperties.java#L92
- https://github.com/spinnaker/igor/blob/b76c323e191b8a8acbe825f8a832ddd3f9460fb5/igor-core/src/main/java/com/netflix/spinnaker/igor/polling/CommonPollingMonitor.java#L177

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.
